variants/audit-logging/services/audit_log.rb

```diff
@@ -0,0 +1,92 @@
+##
+# The log_* class methods are the intended public interface for this class. You
+# should create a method to build each kind of auditable event in your system.
+#
+# The params of the method decide what context is required (and what is
+# optional) for each event (context can and will vary between events).
+#
+# The class method will take the given params, transform them as required and
+# create an instance of AuditLog from them which it will immediately send to the
+# designated log destination.
+#
+# Example usage:
+#
+#   # The "user signed in" event requires details of the current user and the
+#   # remote IP address they signed in from
+#   AuditLog.log_user_sign_in(current_user: current_user, remote_ip: request.remote_ip)
+#
+#   # The "An instance of Foo was created" event needs to know the user who created
+#   # it and the id of the Foo which was created
+#   my_foo = Foo.new(...)
+#   AuditLog.log_create_foo(current_user: current_user, foo_id: my_foo.id)
+#
+class AuditLog
+  EXAMPLE_LABEL = "you_should_delete_this_event_type_when_you_have_real_events".freeze
+  # USER_SIGN_IN_LABEL = "user_sign_in".freeze
+  # CREATE_FOO_LABEL = "create_foo_record".freeze
+
+  ##
+  # An example event which you should delete as soon as you have real events in
+  # the system.
+  #
+  # This event needs to know the user who triggered it, their remote IP and, for
+  # $reasons, the phase of the moon.
+  #
+  def self.log_example_event(current_user:, remote_ip:, phase_of_moon:, timestamp: Time.zone.now)
```

Comment:
how do you feel about using CurrentAttributes for default values here? https://api.rubyonrails.org/classes/ActiveSupport/CurrentAttributes.html

Reply:
I'm not :100: sure I follow. I'm assuming you are suggesting a `Current` model and controller concern to populate it as in the example in the docs you linked.

I'm mildly not in favour if it because it feels like more cognitive overhead (aka "stuff") than is worthwhile for the benefit it brings to a starter template like this. It might be something a real app would refactor towards if it had a lot of audit logged endpoints which were all logging similar stuff.